notebook/mortgage_e2e_gquant/mortgage_gquant_plugins.py

```diff
@@ -0,0 +1,1260 @@
+
+    Configuration:
+        conf: {
+            'delete_dataframes': OPTIONAL. Boolean (True or False). Delete the
+                intermediate mortgage dataframes from which an xgboost dmatrix
+                is created. This is to potentially clear up CPU//GPU memory.
+            'xgb_gpu_params': REQUIRED. Dictionary of xgboost trainer
+                parameters.
+        }
+
+        Example of xgb_gpu_params:
+            xgb_gpu_params = {
+                'nround':            100,
+                'max_depth':         8,
+                'max_leaves':        2 ** 8,
+                'alpha':             0.9,
+                'eta':               0.1,
+                'gamma':             0.1,
+                'learning_rate':     0.1,
+                'subsample':         1,
+                'reg_lambda':        1,
+                'scale_pos_weight':  2,
+                'min_child_weight':  30,
+                'tree_method':       'gpu_hist',
+                'n_gpus':            1,
+                'distributed_dask':  True,
+                'loss':              'ls',
+                # 'objective':         'gpu:reg:linear',
+                'objective':         'reg:squarederror',
+                'max_features':      'auto',
+                'criterion':         'friedman_mse',
+                'grow_policy':       'lossguide',
+                'verbose':           True
+            }
+
+    Inputs:
+        mortgage_feat_df_pandas, delinq_df_pandas = inputs[0]
+        These inputs are provided by MortgageWorkflowRunner.
+
+
+    '''
+    def columns_setup(self):
+        '''
+        '''
+        pass
+
+    def process(self, inputs):
+        import gc  # python standard lib garbage collector
+        import xgboost as xgb
+
+        logger = get_mortgage_plugins_logger()
+
+        mortgage_feat_df_pandas, delinq_df_pandas = inputs[0]
+
+        delete_dataframes = self.conf.get('delete_dataframes')
+        xgb_gpu_params = self.conf['xgb_gpu_params']
+
+        logger.info('JUST BEFORE DMATRIX')
+        print_ram_usage()
+
+        logger.info('CREATING DMATRIX')
+        # DMatrix directly from dataframe requires xgboost from rapidsai:
+        #     https://github.com/rapidsai/xgboost
+        # Convert to DMatrix for XGBoost training.
+        xgb_dmatrix = xgb.DMatrix(mortgage_feat_df_pandas, delinq_df_pandas)
+        # logger.info('XGB_DMATRIX:\n', xgb_dmatrix)
+
+        logger.info('JUST AFTER DMATRIX')
+        print_ram_usage()
+
+        # clear CPU/GPU memory
+        if delete_dataframes:
+            del(mortgage_feat_df_pandas)
+            del(delinq_df_pandas)
+
+        gc.collect()
+
+        logger.info('CLEAR MEMORY JUST BEFORE XGBOOST TRAINING')
+        print_ram_usage()
+
+        logger.info('RUNNING XGBOOST TRAINING')
+
+        # booster object
+        bst = xgb.train(
+            xgb_gpu_params, xgb_dmatrix,
+            num_boost_round=xgb_gpu_params['nround'])
+
+        return bst
```

Comment:
so here we get the xgboost trained model, are you going to use it to do predictions?

Reply:
No, the original example did not do any predictions. I only ported the original example.
<https://github.com/rapidsai/notebooks-extended/blob/master/intermediate_notebooks/E2E/mortgage/mortgage_e2e.ipynb>

Training is kinda incomplete. A complete example then re-loads the model and keeps training on the next set of data. <https://stackoverflow.com/questions/38079853/how-can-i-implement-incremental-training-for-xgboost>

I guess what is the scope of this effort, just to port what is there or make it dynamic enough to scale across nodes, train incrementally if not enough resources to train in one go, etc.

Reply:
agree